File: scripts/rag-health-check.ts
#!/usr/bin/env npx ts-node
/**
 * RAG Health Check - Simple validation suite
 *
 * Validates:
 * 1. Vector search is fast (<2s)
 * 2. Both NCCN and webinar content is retrievable
 * 3. End-to-end query works with acceptable latency
 * 4. Evaluation scores are being stored
 *
 * Run: npx ts-node scripts/rag-health-check.ts
 */

import { createClient } from '@supabase/supabase-js';
import OpenAI from 'openai';

const supabase = createClient(
  process.env.SUPABASE_URL!,
  process.env.SUPABASE_SERVICE_ROLE_KEY!
);

const openai = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });

interface TestResult {
  name: string;
  passed: boolean;
  latencyMs?: number;
  details: string;
}

const results: TestResult[] = [];

// Test queries covering different scenarios
const TEST_QUERIES = [
  { query: 'metastatic prostate cancer treatment options', expectTier: 'tier_1', description: 'NCCN guideline query' },
  { query: 'bipolar androgen therapy BAT', expectTier: 'tier_3', description: 'Webinar-specific query' },
  { query: 'Bob Gatenby adaptive therapy', expectTier: 'tier_3', description: 'Speaker-specific query' },
  { query: 'BRCA mutation breast cancer', expectTier: 'tier_1', description: 'Biomarker query' },
];

const THRESHOLDS = {
  maxSearchLatencyMs: 2000,
  maxE2ELatencyMs: 10000,
  minSimilarityScore: 0.35,
  minChunksReturned: 3,
};

async function generateEmbedding(text: string): Promise<number[]> {
  const response = await openai.embeddings.create({
    model: 'text-embedding-3-small',
    input: text,
  });
  return response.data[0].embedding;
}

async function testVectorSearch(query: string, expectTier: string): Promise<TestResult> {
  const start = Date.now();

  try {
    const embedding = await generateEmbedding(query);
    const embeddingTime = Date.now() - start;

    const searchStart = Date.now();
    const { data, error } = await supabase.rpc('match_chunks', {
      query_embedding: embedding,
      match_count: 10,
      match_threshold: 0.35,
    });
    const searchTime = Date.now() - searchStart;
    const totalTime = Date.now() - start;

    if (error) throw error;

    const chunks = data || [];
    const hasTier = chunks.some((c: any) => c.content_tier === expectTier);
    const avgSimilarity = chunks.length > 0
      ? chunks.reduce((sum: number, c: any) => sum + c.similarity, 0) / chunks.length
      : 0;

    const passed =
      searchTime < THRESHOLDS.maxSearchLatencyMs &&
      chunks.length >= THRESHOLDS.minChunksReturned &&
      avgSimilarity >= THRESHOLDS.minSimilarityScore;

    return {
      name: `Vector Search: "${query.slice(0, 30)}..."`,
      passed,
      latencyMs: searchTime,
      details: `${chunks.length} chunks, avg sim: ${avgSimilarity.toFixed(3)}, has ${expectTier}: ${hasTier}, search: ${searchTime}ms`,
    };
  } catch (err) {
    return {
      name: `Vector Search: "${query.slice(0, 30)}..."`,
      passed: false,
      details: `Error: ${err}`,
    };
  }
}

async function testTierCoverage(): Promise<TestResult> {
  try {
    // Check tier_1 (NCCN) has embeddings
    const { count: tier1Count } = await supabase
      .from('guideline_chunks')
      .select('id', { count: 'exact' })
      .eq('content_tier', 'tier_1')
      .eq('status', 'active')
      .not('chunk_embedding_vec', 'is', null);

    // Check tier_3 (webinars) has embeddings
    const { count: tier3Count } = await supabase
      .from('guideline_chunks')
      .select('id', { count: 'exact' })
      .eq('content_tier', 'tier_3')
      .eq('status', 'active')
      .not('chunk_embedding_vec', 'is', null);

    const passed = (tier1Count || 0) > 1000 && (tier3Count || 0) > 1000;

    return {
      name: 'Tier Coverage',
      passed,
      details: `tier_1 (NCCN): ${tier1Count} chunks, tier_3 (webinars): ${tier3Count} chunks`,
    };
  } catch (err) {
    return {
      name: 'Tier Coverage',
      passed: false,
      details: `Error: ${err}`,
    };
  }
}

async function testIndexExists(): Promise<TestResult> {
  try {
    // Simple test: run a vector query and check it's fast
    const embedding = await generateEmbedding('test query');

    const start = Date.now();
    const { data, error } = await supabase.rpc('match_chunks', {
      query_embedding: embedding,
      match_count: 5,
      match_threshold: 0.3,
    });
    const latency = Date.now() - start;

    // If latency < 2s, index is likely working
    const passed = latency < THRESHOLDS.maxSearchLatencyMs && !error;

    return {
      name: 'Vector Index Health',
      passed,
      latencyMs: latency,
      details: passed
        ? `Index responding in ${latency}ms (threshold: ${THRESHOLDS.maxSearchLatencyMs}ms)`
        : `Slow response: ${latency}ms - index may not be working`,
    };
  } catch (err) {
    return {
      name: 'Vector Index Health',
      passed: false,
      details: `Error: ${err}`,
    };
  }
}

async function testRecentEvaluations(): Promise<TestResult> {
  try {
    // Check if evaluations are being stored (last 24 hours)
    const oneDayAgo = new Date(Date.now() - 24 * 60 * 60 * 1000).toISOString();

    const { data, count } = await supabase
      .from('response_evaluations')
      .select('overall_confidence, total_latency_ms', { count: 'exact' })
      .gte('created_at', oneDayAgo)
      .limit(100);

    const avgConfidence = data && data.length > 0
      ? data.reduce((sum, r) => sum + (r.overall_confidence || 0), 0) / data.length
      : 0;

    const avgLatency = data && data.length > 0
      ? data.reduce((sum, r) => sum + (r.total_latency_ms || 0), 0) / data.length
      : 0;

    return {
      name: 'Evaluation Storage',
      passed: true, // Info only - doesn't fail
      details: `Last 24h: ${count || 0} evals, avg confidence: ${avgConfidence.toFixed(2)}, avg latency: ${Math.round(avgLatency)}ms`,
    };
  } catch (err) {
    return {
      name: 'Evaluation Storage',
      passed: true,
      details: `Could not check: ${err}`,
    };
  }
}

async function testWebinarUrls(): Promise<TestResult> {
  try {
    // Check that webinar URLs point to Supabase storage, not fake domains
    const { data } = await supabase
      .from('guideline_chunks')
      .select('url')
      .eq('content_tier', 'tier_3')
      .eq('status', 'active')
      .limit(100);

    const badUrls = (data || []).filter(
      (r) => r.url && (r.url.includes('leafscience.org') || r.url.includes('example.com'))
    );

    const passed = badUrls.length === 0;

    return {
      name: 'Webinar URL Quality',
      passed,
      details: passed
        ? 'All checked URLs point to valid storage locations'
        : `Found ${badUrls.length} URLs with invalid domains`,
    };
  } catch (err) {
    return {
      name: 'Webinar URL Quality',
      passed: false,
      details: `Error: ${err}`,
    };
  }
}

async function runHealthCheck() {
  console.log('═'.repeat(60));
  console.log('RAG HEALTH CHECK');
  console.log('═'.repeat(60));
  console.log(`Time: ${new Date().toISOString()}\n`);

  // Run all tests
  results.push(await testIndexExists());
  results.push(await testTierCoverage());
  results.push(await testWebinarUrls());

  for (const testCase of TEST_QUERIES) {
    results.push(await testVectorSearch(testCase.query, testCase.expectTier));
  }

  results.push(await testRecentEvaluations());

  // Print results
  console.log('\nRESULTS:');
  console.log('─'.repeat(60));

  let passed = 0;
  let failed = 0;

  for (const result of results) {
    const status = result.passed ? '✅ PASS' : '❌ FAIL';
    const latency = result.latencyMs ? ` (${result.latencyMs}ms)` : '';
    console.log(`${status} ${result.name}${latency}`);
    console.log(`       ${result.details}`);

    if (result.passed) passed++;
    else failed++;
  }

  console.log('\n' + '═'.repeat(60));
  console.log(`SUMMARY: ${passed} passed, ${failed} failed`);
  console.log('═'.repeat(60));

  // Exit with error code if any critical tests failed
  if (failed > 0) {
    process.exit(1);
  }
}

runHealthCheck().catch(console.error);

File: rag_scripts_guidelines/rag_health_check.py
#!/usr/bin/env python3
"""
RAG Health Check - Quick validation suite

Validates:
1. Vector search is fast (<2s)
2. Both NCCN and webinar content is retrievable
3. Evaluation data is being stored
4. URLs are valid

Run: python rag_health_check.py
"""

import os
import sys
import time
from datetime import datetime, timedelta, timezone
from dotenv import load_dotenv
from supabase import create_client
import openai

load_dotenv()

supabase = create_client(
    os.getenv('SUPABASE_URL'),
    os.getenv('SUPABASE_SERVICE_ROLE_KEY')
)
openai.api_key = os.getenv('OPENAI_API_KEY')

# Test configuration
TEST_QUERIES = [
    {'query': 'metastatic prostate cancer treatment options', 'expect_tier': 'tier_1', 'desc': 'NCCN query'},
    {'query': 'bipolar androgen therapy BAT', 'expect_tier': 'tier_3', 'desc': 'Webinar query'},
    {'query': 'Bob Gatenby adaptive therapy', 'expect_tier': 'tier_3', 'desc': 'Speaker query'},
]

THRESHOLDS = {
    'max_search_latency_ms': 2000,
    'min_similarity': 0.35,
    'min_chunks': 3,
}

results = []


def log_result(name: str, passed: bool, details: str, latency_ms: int = None):
    """Log a test result"""
    results.append({
        'name': name,
        'passed': passed,
        'details': details,
        'latency_ms': latency_ms
    })
    status = '✅ PASS' if passed else '❌ FAIL'
    latency_str = f' ({latency_ms}ms)' if latency_ms else ''
    print(f'{status} {name}{latency_str}')
    print(f'       {details}')


def generate_embedding(text: str) -> list:
    """Generate embedding using OpenAI"""
    response = openai.embeddings.create(
        model='text-embedding-3-small',
        input=text
    )
    return response.data[0].embedding


def test_vector_search(query: str, expect_tier: str, desc: str):
    """Test vector search for a query"""
    try:
        # Generate embedding
        embedding = generate_embedding(query)

        # Run search
        start = time.time()
        result = supabase.rpc('match_chunks', {
            'query_embedding': embedding,
            'match_count': 10,
            'match_threshold': 0.35,
        }).execute()
        latency = int((time.time() - start) * 1000)

        chunks = result.data or []
        has_tier = any(c.get('content_tier') == expect_tier for c in chunks)
        avg_sim = sum(c.get('similarity', 0) for c in chunks) / len(chunks) if chunks else 0

        passed = (
            latency < THRESHOLDS['max_search_latency_ms'] and
            len(chunks) >= THRESHOLDS['min_chunks'] and
            avg_sim >= THRESHOLDS['min_similarity']
        )

        log_result(
            f'Search: {desc}',
            passed,
            f'{len(chunks)} chunks, avg sim: {avg_sim:.3f}, has {expect_tier}: {has_tier}',
            latency
        )
    except Exception as e:
        log_result(f'Search: {desc}', False, f'Error: {e}')


def test_tier_coverage():
    """Test that both tiers have embedded chunks"""
    try:
        # Count tier_1 chunks with embeddings
        tier1 = supabase.table('guideline_chunks').select('id', count='exact') \
            .eq('content_tier', 'tier_1').eq('status', 'active') \
            .not_.is_('chunk_embedding_vec', 'null').execute()

        # Count tier_3 chunks with embeddings
        tier3 = supabase.table('guideline_chunks').select('id', count='exact') \
            .eq('content_tier', 'tier_3').eq('status', 'active') \
            .not_.is_('chunk_embedding_vec', 'null').execute()

        tier1_count = tier1.count or 0
        tier3_count = tier3.count or 0

        passed = tier1_count > 1000 and tier3_count > 1000

        log_result(
            'Tier Coverage',
            passed,
            f'tier_1 (NCCN): {tier1_count}, tier_3 (webinars): {tier3_count}'
        )
    except Exception as e:
        log_result('Tier Coverage', False, f'Error: {e}')


def test_index_health():
    """Test that vector index is working (fast response)"""
    try:
        embedding = generate_embedding('test query for index health')

        start = time.time()
        result = supabase.rpc('match_chunks', {
            'query_embedding': embedding,
            'match_count': 5,
            'match_threshold': 0.3,
        }).execute()
        latency = int((time.time() - start) * 1000)

        passed = latency < THRESHOLDS['max_search_latency_ms']

        log_result(
            'Vector Index',
            passed,
            f'Response in {latency}ms (threshold: {THRESHOLDS["max_search_latency_ms"]}ms)',
            latency
        )
    except Exception as e:
        log_result('Vector Index', False, f'Error: {e}')


def test_webinar_urls():
    """Test that webinar URLs are valid (not fake domains)"""
    try:
        result = supabase.table('guideline_chunks').select('url') \
            .eq('content_tier', 'tier_3').eq('status', 'active').limit(100).execute()

        bad_urls = [r for r in (result.data or [])
                    if r.get('url') and ('leafscience.org' in r['url'] or 'example.com' in r['url'])]

        passed = len(bad_urls) == 0

        log_result(
            'Webinar URLs',
            passed,
            'All URLs valid' if passed else f'{len(bad_urls)} URLs with invalid domains'
        )
    except Exception as e:
        log_result('Webinar URLs', False, f'Error: {e}')


def test_recent_evaluations():
    """Check if evaluations are being stored"""
    try:
        one_day_ago = (datetime.now(timezone.utc) - timedelta(days=1)).isoformat()

        result = supabase.table('response_evaluations').select('overall_confidence, total_latency_ms', count='exact') \
            .gte('created_at', one_day_ago).limit(100).execute()

        data = result.data or []
        count = result.count or 0

        avg_conf = sum(r.get('overall_confidence', 0) for r in data) / len(data) if data else 0
        avg_latency = sum(r.get('total_latency_ms', 0) for r in data) / len(data) if data else 0

        log_result(
            'Eval Storage (24h)',
            True,  # Info only
            f'{count} evals, avg confidence: {avg_conf:.2f}, avg latency: {int(avg_latency)}ms'
        )
    except Exception as e:
        log_result('Eval Storage', True, f'Could not check: {e}')


def main():
    print('═' * 60)
    print('RAG HEALTH CHECK')
    print('═' * 60)
    print(f'Time: {datetime.now().isoformat()}\n')

    # Run tests
    test_index_health()
    test_tier_coverage()
    test_webinar_urls()

    for tc in TEST_QUERIES:
        test_vector_search(tc['query'], tc['expect_tier'], tc['desc'])

    test_recent_evaluations()

    # Summary
    passed = sum(1 for r in results if r['passed'])
    failed = sum(1 for r in results if not r['passed'])

    print('\n' + '═' * 60)
    print(f'SUMMARY: {passed} passed, {failed} failed')
    print('═' * 60)

    if failed > 0:
        sys.exit(1)


if __name__ == '__main__':
    main()
